compiler/damlc/daml-doc/src/DA/Daml/Doc/Render/Markdown.hs

```diff
@@ -3,236 +3,80 @@
-      [ "| " <> adjust fLen "Field"   <> " | Type/Description |"
-      , "| :" <> T.replicate (fLen - 1) "-" <> " | :----------------"
-      ]
-
-    fieldRows = renderLinesDep $ \env -> concat
-      [ ("| " <> adjust fLen (escapeMd $ unFieldname fd_name)
-              <> " | " <> type2md env fd_type <> " |")
-        : maybe []
-              (\descr -> [col1Empty <> removeLineBreaks (unDocText descr) <> " |"])
-              fd_descr
-      | FieldDoc{..} <- fds ]
-
-    -- Markdown does not support multi-row cells so we have to remove
-    -- line breaks.
-    removeLineBreaks = T.unwords . T.lines
-
-    fLen = maximum $ 5 : map (T.length . escapeMd . unFieldname . fd_name) fds
-      -- 5 = length of "Field" header
-
-    col1Empty = "| " <> T.replicate fLen " " <> " | "
-
--- | Render a type. Nested type applications are put in parentheses.
-type2md :: RenderEnv -> Type -> T.Text
-type2md env = f 0
-  where
-    -- 0 = no brackets
-    -- 1 = brackets around function
-    -- 2 = brackets around function AND application
-    f :: Int -> Type -> T.Text
-    f _ (TypeApp a n []) = link a n
-    f i (TypeApp a n as) = (if i >= 2 then inParens else id) $
-        T.unwords (link a n : map (f 2) as)
-    f i (TypeFun ts) = (if i >= 1 then inParens else id) $
-        T.intercalate " -> " $ map (f 1) ts
-    f _ (TypeList t1) = "\\[" <> f 0 t1 <> "\\]"
-    f _ (TypeTuple ts) = "(" <> T.intercalate ", " (map (f 0) ts) <>  ")"
-
-    link :: Maybe Anchor -> Typename -> T.Text
-    link Nothing n = escapeMd $ unTypename n
-    link (Just anchor) n =
+renderMd :: RenderEnv -> RenderOut -> [T.Text]
+renderMd env = \case
+    RenderSpaced chunks -> renderMdSpaced env chunks
+    RenderModuleHeader title -> ["# " <> title]
+    RenderSectionHeader title -> ["## " <> title]
+    RenderBlock block -> blockquote (renderMd env block)
+    RenderList items -> spaced (map (bullet . renderMd env) items)
+    RenderFields fields -> renderMdFields env fields
+    RenderPara text -> [renderMdText env text]
+    RenderDocs docText -> T.lines . unDocText $ docText
+    RenderAnchor anchor -> [anchorTag anchor]
+
+renderMdWithAnchor :: RenderEnv -> Anchor -> RenderOut -> [T.Text]
+renderMdWithAnchor env anchor = \case
+    RenderModuleHeader title -> ["# " <> anchorTag anchor <> title]
+    RenderSectionHeader title -> ["## " <> anchorTag anchor <> title]
+    RenderPara text -> [anchorTag anchor <> renderMdText env text]
+    other -> anchorTag anchor : renderMd env other
+
+renderMdSpaced :: RenderEnv -> [RenderOut] -> [T.Text]
+renderMdSpaced env = spaced . renderMds env
+
+renderMds :: RenderEnv -> [RenderOut] -> [[T.Text]]
+renderMds env = \case
+    RenderAnchor anchor : next : rest ->
+        renderMdWithAnchor env anchor next : renderMds env rest
+    next : rest -> renderMd env next : renderMds env rest
+    [] -> []
+
+renderMdText :: RenderEnv -> RenderText -> T.Text
+renderMdText env = \case
+    RenderConcat ts -> mconcatMap (renderMdText env) ts
```

Comment:
Is there any reason why you’re using `mconcatMap` instead of `foldMap` which is even in `base`? This PR does not introduce this but I would generally prefer `foldMap` since I had to check if `mconcatMap` actually does what I thought since I usually don’t use it.

Reply:
I must be weird because I find `mconcatMap` more intuitive as a generalize  `concatMap`. It also has a more specific type than `foldMap`.